pkg/agentconfig/container.go

```diff
@@ -313,6 +332,41 @@ func appendSecretVolume(env dos.Env, annotation, volumeName string, pod *core.Po
 	return volumes
 }
 
+// ConfiguredContainers will find each container in the given config and match it against a container
+// in the pod using its name. The returned slice is guaranteed to use the same index as the Sidecar.Containers slice.
+func ConfiguredContainers(ctx context.Context, pod *core.Pod, config *Sidecar) []*core.Container {
+	cns := pod.Spec.Containers
+	result := make([]*core.Container, len(config.Containers))
+	for ci, cc := range config.Containers {
+		for i := range cns {
+			app := &cns[i]
+			if app.Name == ContainerName {
+				// The pod might hold JSON of replaced containers from an earlier patch
+				annName := ReplacedContainerAnnotationPrefix + cc.Name
+				if appJson, ok := pod.ObjectMeta.Annotations[annName]; ok {
+					var cn core.Container
+					err := json.Unmarshal([]byte(appJson), &cn)
+					if err != nil {
+						dlog.Errorf(ctx, "failed to unmarshal container annotation %s: %v", annName, err)
+					}
+					result[ci] = &cn
+					break
+				}
+			} else if app.Name == cc.Name {
+				result[ci] = app
+				break
+			}
+		}
+	}
+	return result
+}
+
+func eachConfiguredContainer(configureContainers []*core.Container, config *Sidecar, f func(*core.Container, *Container)) {
+	for i, cn := range configureContainers {
+		f(cn, config.Containers[i])
```

Comment:
is it possible that a container will be nil at runtime?
Technically from the perspective of `ConfiguredContainers` function that might be the case if we're unable to find the configured container.
If so, we may want to filter out nil values here as the callbacks we pass to `eachConfiguredContainer` will panic due to them expecting a non-nil value.

Reply:
Good point, A nil entry in the `configContainers` should never be possible, but looking at the `ConfiguredContainers` function, I realize that if the `pod.Spec.Containers` for some reason is out of sync with the `config.Containers`, and no longer contain all the containers that were present at the time when the `config.Containers` was generated, then that could actually happen.